### ISSUE TITLE
Example Notebooks Failing

### ISSUE THREAD

**Issue by jetlime** [IBM/URET#6]
After following the installation guidelines for the notebooks, the installation of URET results in an error after running ```pip install -e .```:

1. Invalid version of sklearn.
```bash
  Installing build dependencies ...
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
INFO: pip is looking at multiple versions of uret to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement scikit-learn==0.22.2 (from uret) (from versions: 0.9, 0.10, 0.11, 0.12, 0.12.1, 0.13, 0.13.1, 0.14, 0.14.1, 0.15.0, 0.15.1, 0.15.2, 0.16.0, 0.16.1, 0.17, 0.17.1, 0.18, 0.18.1, 0.18.2, 0.19.0, 0.19.1, 0.19.2, 0.20.0, 0.20.1, 0.20.2, 0.20.3, 0.20.4, 0.21.1, 0.21.2, 0.21.3, 0.22, 0.22.1, 0.22.2.post1, 0.23.0, 0.23.1, 0.23.2, 0.24.0, 0.24.1, 0.24.2, 1.0, 1.0.1, 1.0.2, 1.1.0, 1.1.1, 1.1.2, 1.1.3, 1.2.0rc1, 1.2.0, 1.2.1, 1.2.2, 1.3.0rc1, 1.3.0, 1.3.1, 1.3.2, 1.4.0rc1, 1.4.0, 1.4.1.post1)
ERROR: No matching distribution found for scikit-learn==0.22.2
```

The ```notebook/setup.py``` file shall define a scikit-learn version of 0.22.2.post1 and not 0.22.2

2. Unsuccessful build:

```bash
         96 |      &                    shiftr, shifti )
            |                                1
      Warning: Unused dummy argument ‘shiftr’ at (1) [-Wunused-dummy-argument]
      stat.h:8:27:
      
      Warning: Unused variable ‘t2’ declared at (1) [-Wunused-variable]
      stat.h:8:31:
      
      Warning: Unused variable ‘t3’ declared at (1) [-Wunused-variable]
      stat.h:8:35:
      
      Warning: Unused variable ‘t4’ declared at (1) [-Wunused-variable]
      stat.h:8:39:
      
      Warning: Unused variable ‘t5’ declared at (1) [-Wunused-variable]
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/zmout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/icnteq.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/ivout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/icopy.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/cvout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssgets.f
      stat.h:8:27:
      
      Warning: Unused variable ‘t2’ declared at (1) [-Wunused-variable]
      stat.h:8:31:
      
      Warning: Unused variable ‘t3’ declared at (1) [-Wunused-variable]
      stat.h:8:35:
      
      Warning: Unused variable ‘t4’ declared at (1) [-Wunused-variable]
      stat.h:8:39:
      
      Warning: Unused variable ‘t5’ declared at (1) [-Wunused-variable]
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/svout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/dmout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/zvout.f
      gfortran:f77: /tmp/pip-install-0pv7armk/scipy_b39ea6bdc3ee4c1ea14e5e0d05ee3820/scipy/_build_utils/src/wrap_dummy_g77_abi.f
      error: Command "/usr/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -fPIC -O3 -funroll-loops -Iscipy/sparse/linalg/eigen/arpack/ARPACK/SRC -I/tmp/pip-build-env-azy5jgtb/overlay/lib64/python3.9/site-packages/numpy/core/include -c -c scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnaup2.f -o build/temp.linux-x86_64-cpython-39/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnaup2.o" failed with exit status 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for scipy
Successfully built uret
Failed to build scikit-learn scipy
ERROR: Could not build wheels for scikit-learn, scipy, which is required to install pyproject.toml-based projects
```

**Comment by jetlime** [comment]
By playing around with the dependencies, I found matching pairs for which I can compile the project. However, I now have the following error on all notebooks when executing the first attack cell (namely e.g https://github.com/IBM/URET/blob/6531041e5e8d9606b91bc0119715d774cb9812b0/notebooks/HMDA_lookup.ipynb#L248C3-L291C5):

```bash
ValueError: node array from the pickle has an incompatible dtype:
- expected: {'names': ['left_child', 'right_child', 'feature', 'threshold', 'impurity', 'n_node_samples', 'weighted_n_node_samples', 'missing_go_to_left'], 'formats': ['<i8', '<i8', '<i8', '<f8', '<f8', '<i8', '<f8', 'u1'], 'offsets': [0, 8, 16, 24, 32, 40, 48, 56], 'itemsize': 64}
- got     : [('left_child', '<i8'), ('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'), ('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples', '<f8')]
```

Is this behavior related to a wrong package version I use, or to a bug present in the notebook's python code?

**Comment by keykholt** [comment]
I was able to resolve the issue by using python 3.8. I'm not sure why it doesn't work with older python versions. I've updated the notebook's setup file as there was an additional package I needed to downgrade in order to run jupyter.

Steps that work for me (assuming I'm using anaconda):
1. `conda create -n URET python==3.8`
2. `cp notebooks/setup.py .`
3. `pip install -e .`


I assume the other issue you had is a result of the changed dependencies. Probably the model and the version of the scikit package.

Let me know if you still have issues.
